### PR TITLE
TRT-932: Adding alert query for known alerts

### PR DESCRIPTION
**Pull request by deepsm007**
Updating ci-tools BigQuery query to include first time alert nurp was observed and updates the query_results.json

### CODE REVIEW

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go [hunk 1]
```diff
@@ -17,6 +18,13 @@ type HistoricalData interface {
 	GetJobRuns() int
 }
 
+type HistoricalDataWithSetters interface {
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go:21]: I don't think we need this

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go [hunk 2]
```diff
@@ -33,7 +41,10 @@ type AlertHistoricalDataRow struct {
 	AlertName      string
 	AlertNamespace string
 	AlertLevel     string
+	FirstObserved  time.Time
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go:44]: There were only needed for alerts so let's move there

File: pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go [hunk 3]
```diff
@@ -52,7 +52,7 @@ type JobLister interface {
 
 type HistoricalDataClient interface {
 	ListDisruptionHistoricalData(ctx context.Context) ([]jobrunaggregatorapi.HistoricalData, error)
-	ListAlertHistoricalData(ctx context.Context) ([]jobrunaggregatorapi.HistoricalData, error)
+	ListAlertHistoricalData(ctx context.Context) ([]jobrunaggregatorapi.HistoricalDataWithSetters, error)
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go:55]: How about returning  []*jobrunaggregatorapi.AlertHistoricalDataRow

File: pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go [hunk 4]
```diff
@@ -168,7 +168,7 @@ ORDER BY
 	return jobrunaggregatorapi.ConvertToHistoricalData(disruptionDataSet), nil
 }
 
-func (c *ciDataClient) ListAlertHistoricalData(ctx context.Context) ([]jobrunaggregatorapi.HistoricalData, error) {
+func (c *ciDataClient) ListAlertHistoricalData(ctx context.Context) ([]jobrunaggregatorapi.HistoricalDataWithSetters, error) {
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go:171]: Don't call jobrunaggregatorapi.ConvertToHistoricalData(alertDataSet), nil below just return  []*jobrunaggregatorapi.AlertHistoricalDataRow when done

File: pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go [hunk 5]
```diff
@@ -91,6 +91,30 @@ func (o *JobRunHistoricalDataAnalyzerOptions) Run(ctx context.Context) error {
 	return nil
 }
 
+func (o *JobRunHistoricalDataAnalyzerOptions) getAlertData(ctx context.Context) []jobrunaggregatorapi.HistoricalData {
+	var allKnownAlerts []*jobrunaggregatorapi.KnownAlertRow
+	var newHistoricalData []jobrunaggregatorapi.HistoricalDataWithSetters
+
+	newHistoricalData, err := o.ciDataClient.ListAlertHistoricalData(ctx)
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go:98]: This should now be  []*jobrunaggregatorapi.AlertHistoricalDataRow and you can use it in your innner loop below

File: pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go [hunk 6]
```diff
@@ -91,6 +91,30 @@ func (o *JobRunHistoricalDataAnalyzerOptions) Run(ctx context.Context) error {
...
+		return nil
+	}
+	allKnownAlerts, err = o.ciDataClient.ListAllKnownAlerts(ctx)
+	if err != nil {
+		return nil
+	}
+	for _, alerts := range allKnownAlerts {
+		for _, jobData := range newHistoricalData {
+			if alerts.AlertName == jobData.GetName() {
+				jobData.SetFirstObserved(alerts.FirstObserved)
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go:109]: you don't need the setters if you are using AlertHistoricalDataRow

File: pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go [hunk 7]
```diff
@@ -91,6 +91,30 @@ func (o *JobRunHistoricalDataAnalyzerOptions) Run(ctx context.Context) error {
...
+				jobData.SetLastObserved(alerts.LastObserved)
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go:110]: break out of the inner loop if you found the one you want

File: pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go [hunk 8]
```diff
@@ -91,6 +91,30 @@ func (o *JobRunHistoricalDataAnalyzerOptions) Run(ctx context.Context) error {
...
+			}
+		}
+	}
+
+	return newHistoricalData
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go:115]: return jobrunaggregatorapi.ConvertToHistoricalData(newHistoricalData),

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go (hunk 2 above)
Comment (deepsm007) [reply pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go:37]: they are in alerts

**Comment by DennisPeriquet** [comment]
I see this in the query_resuilts.json uploaded to the Jira:

```
...
  "FirstObserved": "0001-01-01T00:00:00Z",
  "LastObserved": "0001-01-01T00:00:00Z",
...
```

**Comment by DennisPeriquet** [comment]
the date looks like 0s
/hold

**Comment by deepsm007** [comment]
Updated the code now it’s generating correct data in the file, new file is in the jira

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go [hunk 9]
```diff
@@ -44,7 +47,7 @@ func (a *AlertHistoricalDataRow) GetJobData() HistoricalJobData {
 	return a.HistoricalJobData
 }
 func (a *AlertHistoricalDataRow) GetName() string {
-	return fmt.Sprintf("%s %s %s", a.AlertName, a.AlertNamespace, a.AlertLevel)
+	return a.AlertName
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go:50]: Seems like we are loosing context here.  So if an alert with the same name but different namespace exists we can no longer tell the difference based on the name.  Looks like this is only used in a markdown table currently did you really need this change?

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go [hunk 10]
```diff
@@ -72,6 +75,8 @@ func (a *AlertHistoricalDataRow) GetKey() string {
 		a.Platform,
 		a.Network,
 		a.Topology,
+		a.FirstObserved,
+		a.LastObserved,
```
Comment (neisw) [inline pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go:79]: Do we want these in the keys?  LastObserved will be changing over time, if the keys in the maps that are generated change won't that break https://github.com/deepsm007/ci-tools/blob/33f9a92cb9fbd66e3759f084dbe7ef0f5074c57a/pkg/jobrunaggregator/jobrunhistoricaldataanalyzer/analyzer.go#L162

Have you tried running the update with an existing file and validated things still match up?

File: pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go (hunk 9 above)
Comment (deepsm007) [reply pkg/jobrunaggregator/jobrunaggregatorapi/types_row_historical_data.go]: reverted back to original

**Pull request merged**